### PR TITLE
documentation: package.mk

### PR DESCRIPTION
**Pull request by InuSasha**
Since the last changes on the build system, we need an update to our documentation.
I prefer, to place it in github  instead of wiki, because it is easier to find for developers (as special *new* developers)
After finishing/merging this, we should link the wiki to this location.

Actual it is"Work in progress"
Open Thinks, which has to be done first:
- [x] description for `PKG_NEEDS_UNPACK`
- [x] functions
- [x] spelling check 
- [x] 2 `package.mk` templates (for non-addon and addon)
       incl. how to add a new package
- [x] mega-squash :) (i had to look, how github parse the markdown)

I copied most from wiki and the readme in packages/addons

### CODE REVIEW

**Comment by awiouy** [comment]
PKG_LONGDESC and PKG_SHORTDESC are only used for addons, to populate `addon.xml` file.
However, all packages must have PKG_LONGDESC, for documentation.

In summary:
- addons must have both PKG_LONGDESC and PKG_SHORTDESC
- other packages must have PKG_LONGDESC

**Comment by InuSasha** [comment]
so PKG_SHORTDESC is only required for addons?

**Comment by InuSasha** [comment]
I added part "Add a new package to the Image"
and two templates for the `package.mk`.
(The package.mk.addon_template replaces the README.md from packages/addons.
Before one is asking, we have two addon location under packages/, now: addons and mediacenter/kodi-binary-addons)

Now, i put down my pencil and wait for responses.

The variables needs a description: 
* PKG_DISCLAIMER
* PKG_NEED_UNPACK
* PKG_ADDON_IS_STANDALONE

**Comment by InuSasha** [comment]
* `PKG_DISCLAIMER` and `PKG_ADDON_IS_STANDALONE` described.
* all squashed

File: packages/packages.mk.addon_template [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="[package name]"
+PKG_VERSION="[package upstream version]"
```
Comment (chewitt) [inline packages/packages.mk.addon_template]: "[package version identifier or unique githash]"

File: packages/packages.mk.addon_template [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+PKG_REV="100"
+PKG_SHA256="[sha256 hash of the source file, downloaded from PKG_URL]"
+PKG_ARCH="any"
+PKG_LICENSE="[license of the upstream project]"
+PKG_SITE="[url to the upstream project, e.g. http://exmaple.com/libexample"
+PKG_URL="[download url, e.g. https://github.com/example/libexample/archive/$PKG_VERSION.tar.gz]"
+# PKG_MAINTAINER="John Doe (email)" # if you want to be know as maintainer for a addon
+PKG_DEPENDS_TARGET="[needed packages for build, e.g. toolchain zlib openssl]"
```
Comment (chewitt) [inline packages/packages.mk.addon_template]: "[build system dependencies, e.g. toolchain zlib openssl]"

File: packages/packages.mk.addon_template [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+PKG_SECTION="[location under packages, e.g. database]"
+PKG_SHORTDESC="[short description (mostly find at the upstrem project), e.g. libexample: library to calculate examples]"
```
Comment (chewitt) [inline packages/packages.mk.addon_template]: "[short description of the package, e.g. libexample: library to calculate examples]"

File: packages/packages.mk.addon_template [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+PKG_LONGDESC="[long description (mostly find at the upstrem project), e.g. libexample: this project is created for case x and y, to do it, in a best way...]"
```
Comment (chewitt) [inline packages/packages.mk.addon_template]: "[long description of the package, often taken from the package/project website, e.g. libexample: this project is created to calculate examples for x and y, with maximum efficiency and fewer errors]"

File: packages/packages.mk.addon_template [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+PKG_NAME="[package name]"
+PKG_VERSION="[package upstream version]"
+PKG_REV="100"
+PKG_SHA256="[sha256 hash of the source file, downloaded from PKG_URL]"
+PKG_ARCH="any"
+PKG_LICENSE="[license of the upstream project]"
+PKG_SITE="[url to the upstream project, e.g. http://exmaple.com/libexample"
+PKG_URL="[download url, e.g. https://github.com/example/libexample/archive/$PKG_VERSION.tar.gz]"
+# PKG_MAINTAINER="John Doe (email)" # if you want to be know as maintainer for a addon
```
Comment (chewitt) [inline packages/packages.mk.addon_template]: "John Doe" # Full name or forum/GitHub nickname, if you want to be identified as the addon maintainer

^ we should stop using emails, it results in text-length issues in the Kodi GUI and it's just a bad idea to have emails collected these days (privacy, GDPR, etc.)

File: packages/readme.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
+# Structure of package.mk files
+
+## Introduction
+
+The package.mk file defines variables and functions to build a package.
+
+## Variables
+All avialible variable, to control the build behavoir of your package.
+Please use these in the order, listed here.
```
Comment (chewitt) [inline packages/readme.md]: To control the build behaviour of your package, use variables in the top-down order listed here.

File: packages/readme.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+#### Base
+
+| Variable    | Default | Required |Description |
+|-------------|---------|----------|------------|
+| PKG_NAME    | -       | yes | Name of the packaged software application. Should be lowercase |
+| PKG_VERSION | -       | yes | Version of the packaged software application |
+| PKG_SHA256  | -       | yes | SHA256 hashsum of the application download file |
+| PKG_ARCH    | any     | no  | Architectures for which the package builds. `any` or a space separated list of `aarch64`, `arm` or `x86_64` |
+| PKG_LICENSE | -       | yes | License of the software application. [Reference](https://github.com/LibreELEC/LibreELEC.tv/tree/master/licenses) |
```
Comment (chewitt) [inline packages/readme.md]: It needs to be agreed in the team, but I think we should use SPDX notation for this

File: packages/readme.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_SITE    | -       | yes | Site of the software application |
+| PKG_URL     | -       | yes | Address at which the source of the software application can be retrieved |
+| PKG_MAINTAINER | -    | no  | Your name |
+| PKG_DEPENDS_BOOTSTRAP<br>PKG_DEPENDS_HOST   PKG_DEPENDS_INIT   PKG_DEPENDS_TARGET | - | no | A space separated list of name of packages required to build the software application |
+| PKG_SECTION | -       | no  | virtual if the package only defines dependencies |
+| PKG_SHORTDESC | -     | no<br>yes&nbsp;for&nbsp;addons | Short description of the application software used in various parts of Kodi |
```
Comment (chewitt) [inline packages/readme.md]: "Short description of the software package"

File: packages/readme.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_LONGDESC | -      | yes | Long description of the application software used in various parts of Kodi |
```
Comment (chewitt) [inline packages/readme.md]: "Long description of the package including purpose or function within LibreELEC or Kodi"

File: packages/readme.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+#### Universal Build Option
+| Variable    | Default | Required |Description |
+|-------------|---------|----------|------------|
+| PKG_SOURCE_DIR | -    | no  | Name of the folder to which the source of the software application unpacks. Should only be specified if the source of the software application unpacks to a folder whose name does not start with `$PKG_NAME` |
```
Comment (chewitt) [inline packages/readme.md]: Force the folder name that application sources are unpacked to. Used when sources do not automatically unpack to a folder with the `$PKG_NAME-$PKG_VERSION` naming convention.

File: packages/readme.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_SOURCE_NAME | -   | no  | Name of the file of the source of the software application. Should only be specified if the source of the software application is not the basename of PKG_URL |
```
Comment (chewitt) [inline packages/readme.md]: Force the filename of the application sources. Used when the filename is not the basename of `PKG_URL`

File: packages/readme.md [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_PATCH_DIRS | -    | no  | Change the path for pachtes, to apply to source. Normally the path `./patches` is used. Is this variable is set, the normal path will extended with the value of this variable as a subdirectory: `./patches/${PKG_PATCH_DIRS}` |
```
Comment (chewitt) [inline packages/readme.md]: Patches in `./patches` are automatically applied after package unpack. Use this option to include patches from an additional folder, e.g. `./patches/$PKG_PATCH_DIRS`

File: packages/readme.md [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_NEED_UNPACK | -   | no  | ??? can anyone explane, i have no idea. |
+| PKG_TOOLCHAIN | auto  | no  | Control which of the build toolchains is used. For detailed information, see the [Reference](#toolchain-options). |
+| PKG_BUILD_FLAGS | -   | no  | A Space seperated list of flags, which control often used build modification. Flags can be enabled or disables with a prefixed `+`/`-`. For detailed information, see the [Reference](#build_flags-options). |
+| PKG_PYTHON_VERSION | python2.7 | no | Defines the Python version, which should use. |
```
Comment (chewitt) [inline packages/readme.md]: Define the Python version to be used

File: packages/readme.md [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_IS_KERNEL_PKG | - | no  | Set it to `yes`, for packages which include linux kernel modules |
```
Comment (chewitt) [inline packages/readme.md]: Set to `yes` for packages that include Linux kernel modules

File: packages/readme.md [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+#### Make Options
+| Variable    | Default | Required |Description |
+|-------------|---------|----------|------------|
+| PKG_MAKE_OPTS<br>PKG_MAKE_OPTS_BOOTSTRP<br>PKG_MAKE_OPTS_HOST<br>PKG_MAKE_OPTS_INIT<br> PKG_MAKE_OPTS_TARGET | - | no | Options directly passed to make in the build step
+| PKG_MAKEINSTALL_OPTS_HOST<br>PKG_MAKEINSTALL_OPTS_TARGET | - | no | Options directly passed to make in the install step
+
+#### Addons
+These options only needed, when the package is build as an addon.
+"Required" column is focus addon packages, only. When the package is no addon, none of these options is required.
```
Comment (chewitt) [inline packages/readme.md]: Additional options used when the package builds an addon.

^ only this line is needed

File: packages/readme.md [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_SOURCE_DIR | -    | no  | Name of the folder to which the source of the software application unpacks. Should only be specified if the source of the software application unpacks to a folder whose name does not start with `$PKG_NAME` |
+| PKG_SOURCE_NAME | -   | no  | Name of the file of the source of the software application. Should only be specified if the source of the software application is not the basename of PKG_URL |
+| PKG_PATCH_DIRS | -    | no  | Change the path for pachtes, to apply to source. Normally the path `./patches` is used. Is this variable is set, the normal path will extended with the value of this variable as a subdirectory: `./patches/${PKG_PATCH_DIRS}` |
+| PKG_NEED_UNPACK | -   | no  | ??? can anyone explane, i have no idea. |
+| PKG_TOOLCHAIN | auto  | no  | Control which of the build toolchains is used. For detailed information, see the [Reference](#toolchain-options). |
+| PKG_BUILD_FLAGS | -   | no  | A Space seperated list of flags, which control often used build modification. Flags can be enabled or disables with a prefixed `+`/`-`. For detailed information, see the [Reference](#build_flags-options). |
```
Comment (MilhouseVH) [inline packages/readme.md]: s/seperated/separated (also repeated in line 117)
", which control often used build modification." - maybe instead: "with which to fine-tune the build process."?
s/disables/disabled - "Flags can be enabled or disabled with a `+` or `-` prefix." may also read a little more easily.

File: packages/readme.md [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_SHORTDESC | -     | no<br>yes&nbsp;for&nbsp;addons | Short description of the application software used in various parts of Kodi |
+| PKG_LONGDESC | -      | yes | Long description of the application software used in various parts of Kodi |
+
+#### Universal Build Option
+| Variable    | Default | Required |Description |
+|-------------|---------|----------|------------|
+| PKG_SOURCE_DIR | -    | no  | Name of the folder to which the source of the software application unpacks. Should only be specified if the source of the software application unpacks to a folder whose name does not start with `$PKG_NAME` |
+| PKG_SOURCE_NAME | -   | no  | Name of the file of the source of the software application. Should only be specified if the source of the software application is not the basename of PKG_URL |
+| PKG_PATCH_DIRS | -    | no  | Change the path for pachtes, to apply to source. Normally the path `./patches` is used. Is this variable is set, the normal path will extended with the value of this variable as a subdirectory: `./patches/${PKG_PATCH_DIRS}` |
+| PKG_NEED_UNPACK | -   | no  | ??? can anyone explane, i have no idea. |
```
Comment (chewitt) [inline packages/readme.md]: Space separated list of files or folders to include in package stamp calculation. If the stamp is invalidated through changes to package files or dependent files/folders the package is cleaned and rebuilt. e.g.  `PKG_NEED_UNPACK="$(get_pkg_directory linux)"` will trigger clean/rebuild of a Linux kernel driver package when a change to the `linux` kernel package is detected.

File: packages/readme.md [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_MAKEINSTALL_OPTS_HOST<br>PKG_MAKEINSTALL_OPTS_TARGET | - | no | Options directly passed to make in the install step
+
+#### Addons
+These options only needed, when the package is build as an addon.
+"Required" column is focus addon packages, only. When the package is no addon, none of these options is required.
+
+| Variable    | Default | Required |Description |
+|-------------|---------|----------|------------|
+| PKG_REV     | -       | yes      | The revision number of the addon. Increase on every version. Currently starts at `100`. Please place this variable under `PKG_VERSION` |
+| PKG_IS_ADDON | no     | yes      | Have to set to `yes` |
```
Comment (chewitt) [inline packages/readme.md]: Must be set to `yes`

File: packages/readme.md [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+| PKG_MAKE_OPTS<br>PKG_MAKE_OPTS_BOOTSTRP<br>PKG_MAKE_OPTS_HOST<br>PKG_MAKE_OPTS_INIT<br> PKG_MAKE_OPTS_TARGET | - | no | Options directly passed to make in the build step
+| PKG_MAKEINSTALL_OPTS_HOST<br>PKG_MAKEINSTALL_OPTS_TARGET | - | no | Options directly passed to make in the install step
+
+#### Addons
+These options only needed, when the package is build as an addon.
+"Required" column is focus addon packages, only. When the package is no addon, none of these options is required.
+
+| Variable    | Default | Required |Description |
+|-------------|---------|----------|------------|
+| PKG_REV     | -       | yes      | The revision number of the addon. Increase on every version. Currently starts at `100`. Please place this variable under `PKG_VERSION` |
```
Comment (chewitt) [inline packages/readme.md]: The revision number of the addon (starts at 100). Must be placed after PKG_VERSION. Must be incremented for each new version else Kodi clients will not detect version change and download the updated addon.

**Comment by InuSasha** [comment]
merged your fixes. many thanks @chewitt

File: packages/packages.mk.addon_template [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="[package name]"
+PKG_VERSION="[package version identifier or unique githash]"
+PKG_REV="100"
+PKG_SHA256="[sha256 hash of the source file, downloaded from PKG_URL]"
+PKG_ARCH="any"
+PKG_LICENSE="[license of the upstream project]"
+PKG_SITE="[url to the upstream project, e.g. http://exmaple.com/libexample"
```
Comment (jernejsk) [inline packages/packages.mk.addon_template]: exmaple.com is misspelled

**Comment by InuSasha** [comment]
a "good to go" from me.

**Pull request merged**